File: leetcode/algorithms/111_minimum_depth_of_binary_tree/main.c
#include <stdlib.h>

struct TreeNode {
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
};

/**
 * Recursion: DFS
 * 
 * Complexities:
 *   N - Number of nodes in `root`
 *   H - Height of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(H)
 */
int minDepth(struct TreeNode* root) {
    if (!root) {
        return 0;
    }

    int left_depth = minDepth(root->left);
    int right_depth = minDepth(root->right);

    if (!left_depth || !right_depth) {
        return (left_depth > right_depth ? left_depth : right_depth) + 1;
    }

    return (left_depth < right_depth ? left_depth : right_depth) + 1;
}


// Solution
/**
 * Recursion: DFS
 * 
 * Complexities:
 *   N - Number of nodes in `root`
 *   H - Height of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(H)
 */
 // 두 수 중 작은 값을 구하는 헬퍼 함수
int min(int a, int b) {
    return (a < b) ? a : b;
}

int minDepth(struct TreeNode* root) {
    if (root == NULL) {
        return 0;
    }

    if (root->left == NULL && root->right == NULL) {
        return 1;
    }

    if (root->left == NULL) {
        return 1 + minDepth(root->right);
    }
    if (root->right == NULL) {
        return 1 + minDepth(root->left);
    }

    return 1 + min(minDepth(root->left), minDepth(root->right));
}

File: leetcode/algorithms/111_minimum_depth_of_binary_tree/main.dart
import 'dart:collection';
import 'dart:math';

class TreeNode {
  int val;
  TreeNode? left;
  TreeNode? right;
  TreeNode([this.val = 0, this.left, this.right]);
}

class MinimumDepthOfBinaryTree {
  /**
   * Recursion: DFS
   * 
   * Complexities:
   *   N - Number of nodes in `root`
   *   H - Height of `root`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(H)
   */
  int minDepth(TreeNode? root) {
    if (root == null) {
      return 0;
    }

    int leftDepth = minDepth(root.left);
    int rightDepth = minDepth(root.right);

    if (leftDepth == 0 || rightDepth == 0) {
      return max(leftDepth, rightDepth) + 1;
    }

    return min(leftDepth, rightDepth) + 1;
  }


  // Solution
  /**
   * Solution 1
   * 
   * Iteration: BFS (Queue)
   *
   * Complexities:
   *   N - Number of nodes in `root`
   *   W - Width of `root`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(W)
   */
  int solution1(TreeNode? root) {
    if (root == null) {
      return 0;
    }

    Queue<TreeNode> queue = Queue();
    queue.add(root);
    int depth = 1;

    while (queue.isNotEmpty) {
      int levelSize = queue.length;

      for (int i = 0; i < levelSize; i++) {
        TreeNode current = queue.removeFirst();

        if (current.left == null && current.right == null) {
          return depth;
        }

        if (current.left != null) {
          queue.add(current.left!);
        }
        if (current.right != null) {
          queue.add(current.right!);
        }
      }

      depth++;
    }

    return depth;
  }

  /**
   * Solution 2
   * 
   * Recursion: DFS
   * 
   * Complexities:
   *   N - Number of nodes in `root`
   *   H - Height of `root`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(H)
   */
  int solution2(TreeNode? root) {
    if (root == null) {
      return 0;
    }

    if (root.left == null) {
      return solution2(root.right) + 1;
    }
    if (root.right == null) {
      return solution2(root.left) + 1;
    }

    return min(solution2(root.left), solution2(root.right)) + 1;
  }
}
